agent/consul/catalog_endpoint.go

```diff
@@ -97,6 +99,22 @@ func (c *Catalog) Register(args *structs.RegisterRequest, reply *struct{}) error
 		}
 	}
 
+	if c.srv.config.EnforceValidDNS {
+		dnsNameRe := regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)
```

Comment:
This incurs the cost of recompiling the same regex over and over for every register call, instead dnsNameRegex should be compiled once into an unexported var and reused here.  According to the docs it is safe for concurrent use as well https://golang.org/pkg/regexp/#Regexp